lucet-runtime/lucet-runtime-macros/src/lib.rs

```diff
@@ -0,0 +1,122 @@
+/// directly. Otherwise the behavior of instance termination and timeouts are
+/// undefined. Additionally, the attribute makes the resulting function `unsafe extern "C"`
+/// regardless of how the function is defined, as this ABI is required for all hostcalls.
+///
+/// In most cases, you will want to also provide the `#[no_mangle]` attribute and `pub` visibility
+/// in order for the hostcall to be exported from the final executable.
+///
+/// ```ignore
+/// #[lucet_hostcall]
+/// #[no_mangle]
+/// pub fn yield_5(vmctx: &mut Vmctx) {
+///     vmctx.yield_val(5);
+/// }
+/// ```
+///
+/// Note that `lucet-runtime` must be a dependency of any crate where this attribute is used, and it
+/// may not be renamed (this restriction may be lifted once [this
+/// issue](https://github.com/rust-lang/rust/issues/54363) is resolved).
+#[proc_macro_attribute]
+pub fn lucet_hostcall(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    // determine whether we need to import from `lucet_runtime_internals`; this is useful if we want
+    // to define a hostcall for a target (or tests, more concretely) that doesn't depend on
+    // `lucet-runtime`
+    let in_internals = std::env::var("CARGO_PKG_NAME").unwrap() == "lucet-runtime-internals";
+
+    let mut hostcall = syn::parse_macro_input!(item as syn::ItemFn);
+    let hostcall_ident = hostcall.sig.ident.clone();
+
+    // use the same attributes and visibility as the impl hostcall
+    let attrs = hostcall.attrs.clone();
+    let vis = hostcall.vis.clone();
+
+    // remove #[no_mangle] from the attributes of the impl hostcall if it's there
+    hostcall.attrs = attrs
+        .iter()
+        .filter(|attr| !attr.path.is_ident("no_mangle"))
+        .cloned()
+        .collect();
+    // make the impl hostcall private
+    hostcall.vis = syn::Visibility::Inherited;
+
+    // modify the type signature of the exported raw hostcall based on the original signature
+    let mut raw_sig = hostcall.sig.clone();
+
+    // hostcalls are always unsafe
+    raw_sig.unsafety = Some(syn::Token![unsafe](raw_sig.span()));
+
+    // hostcalls are always extern "C"
+    raw_sig.abi = Some(syn::parse_quote!(extern "C"));
+
+    let vmctx_mod = if in_internals {
+        quote! { lucet_runtime_internals::vmctx }
+    } else {
+        quote! { lucet_runtime::vmctx }
+    };
+
+    // replace the first argument to the raw hostcall with the vmctx pointer
+    if let Some(arg0) = raw_sig.inputs.iter_mut().nth(0) {
+        let lucet_vmctx: syn::FnArg = syn::parse_quote!(vmctx_raw: *mut #vmctx_mod::lucet_vmctx);
+        *arg0 = lucet_vmctx;
+    }
+
+    // the args after the first to provide to the hostcall impl
+    let impl_args = hostcall
+        .sig
+        .inputs
+        .iter()
+        .skip(1)
+        .map(|arg| match arg {
+            syn::FnArg::Receiver(_) => {
+                // this case is an error, but humor the token stream so it will produce a nicer
```

Comment:
This part it a little opaque to me, maybe elaborating on what you mean by 'humor' in the comment could help? This is a lightly held opinion, fwiw.